pkg/karmadactl/util/check.go

```diff
@@ -6,72 +6,138 @@ import (
-			lastErr = err
-			return false, nil
-		}
-		if s.Generation != s.Status.ObservedGeneration {
-			lastErr = fmt.Errorf("expected generation %d, observed generation: %d",
-				s.Generation, s.Status.ObservedGeneration)
-			return false, nil
-		}
-		if (s.Spec.Replicas != nil) && (s.Status.UpdatedReplicas < *s.Spec.Replicas) {
-			lastErr = fmt.Errorf("expected %d replicas, got %d updated replicas",
-				*s.Spec.Replicas, s.Status.UpdatedReplicas)
-			return false, nil
-		}
-		if s.Status.AvailableReplicas < s.Status.UpdatedReplicas {
-			lastErr = fmt.Errorf("expected %d replicas, got %d available replicas",
-				s.Status.UpdatedReplicas, s.Status.AvailableReplicas)
-			return false, nil
+func WaitForStatefulSetRollout(c kubernetes.Interface, sts *appsv1.StatefulSet, timeoutSeconds time.Duration) error {
+	stsWatcher, err := c.AppsV1().StatefulSets(sts.GetNamespace()).Watch(context.Background(), metav1.SingleObject(sts.ObjectMeta))
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-time.After(timeoutSeconds * time.Second):
+			klog.Errorf("wait for StatefulSets(%s/%s) rollout: timeout", sts.GetNamespace(), sts.GetName())
+			return nil
+		case event, ok := <-stsWatcher.ResultChan():
+			if !ok {
+				return fmt.Errorf("wait for Statefulset(%s/%s) rollout: channel broken", sts.GetNamespace(), sts.GetName())
+			}
+
+			status, err := stsStatus(event.Object)
+			if err != nil {
+				return err
+			} else if status == StatusReady {
+				return nil
+			}
 		}
-		return true, nil
-	})
-	if pollError != nil {
-		return fmt.Errorf("wait for Statefulset(%s/%s) rollout: %v: %v", sts.GetNamespace(), sts.GetName(), pollError, lastErr)
 	}
-	return nil
 }
 
 // WaitForDeploymentRollout  wait for Deployment reaches the ready state or timeout.
-func WaitForDeploymentRollout(c kubernetes.Interface, dep *appsv1.Deployment, timeoutSeconds int) error {
-	var lastErr error
-	pollError := wait.PollImmediate(time.Second, time.Duration(timeoutSeconds)*time.Second, func() (bool, error) {
-		d, err := c.AppsV1().Deployments(dep.GetNamespace()).Get(context.TODO(), dep.GetName(), metav1.GetOptions{})
-		if err != nil {
-			lastErr = err
-			return false, nil
-		}
-		if d.Generation != d.Status.ObservedGeneration {
-			lastErr = fmt.Errorf("current generation %d, observed generation %d",
-				d.Generation, d.Status.ObservedGeneration)
-			return false, nil
-		}
-		if (d.Spec.Replicas != nil) && (d.Status.UpdatedReplicas < *d.Spec.Replicas) {
-			lastErr = fmt.Errorf("the number of pods targeted by the deployment (%d pods) is different "+
-				"from the number of pods targeted by the deployment that have the desired template spec (%d pods)",
-				*d.Spec.Replicas, d.Status.UpdatedReplicas)
-			return false, nil
+func WaitForDeploymentRollout(c kubernetes.Interface, dep *appsv1.Deployment, timeoutSeconds time.Duration) error {
+	depWatcher, err := c.AppsV1().Deployments(dep.GetNamespace()).Watch(context.Background(), metav1.SingleObject(dep.ObjectMeta))
```

Comment:
It forgot to call `.Stop()` function. A better choice is  using `k8s.io/client-go/tools/watch` which provides some helper funcs.

Reply:
ok, got it